File: src/main/java/com/exactpro/th2/act/convertors/ConvertorsResponse.java
package com.exactpro.th2.act.convertors;

import com.exactpro.th2.act.grpc.*;
import com.exactpro.th2.common.grpc.Message;
import com.exactpro.th2.common.grpc.Value;
import com.google.protobuf.TextFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

public class ConvertorsResponse {
    private static final Logger LOGGER = LoggerFactory.getLogger(ConvertorsResponse.class);

    public ResponseMessageTyped createResponseMessage(Message message) {
        ResponseMessageTyped responseMessageTyped;
        switch (message.getMetadata().getMessageType()) {
            case "BusinessMessageReject":
                responseMessageTyped = createBusinessMessageReject(message);
                break;
            case "ExecutionReport":
                responseMessageTyped = createExecutionReport(message);
                break;
            case "QuoteStatusReport":
                responseMessageTyped = createQuoteStatusReport(message);
                break;
            case "Quote":
                responseMessageTyped = createQuote(message);
                break;
            case "QuoteAck":
                responseMessageTyped = createQuoteAck(message);
                break;
            case "OrderMassCancelReport":
                responseMessageTyped = createOrderMassCancelReport(message);
                break;
            case "MassQuoteAcknowledgement":
                responseMessageTyped = createMassQuoteAcknowledgement(message);
                break;
            default:
                responseMessageTyped = null;
                break;
        }
        return responseMessageTyped;
    }

    private ResponseMessageTyped createBusinessMessageReject(Message message) {
        BusinessMessageReject businessMessageReject = BusinessMessageReject.newBuilder()
                .setRefMsgType(getField(message, "RefMsgType"))
                .setBusinessRejectReason(Integer.parseInt(getField(message, "BusinessRejectReason")))
                .setBusinessRejectRefId(getField(message, "BusinessRejectRefID"))
                .setRefSeqNum(Integer.parseInt(getField(message, "RefSeqNum")))
                .setText(getField(message, "Text")).build();

        return ResponseMessageTyped.newBuilder().setBusinessMessageReject(businessMessageReject).build();
    }

    private ResponseMessageTyped createExecutionReport(Message message) {
        Message headerField = message.getFieldsMap().get("Header").getMessageValue();

        Header header = Header.newBuilder()
                .setBeginString(getField(headerField, "BeginString"))
                .setSenderCompId(getField(headerField, "SenderCompId"))
                .setSendingTime(getField(headerField, "SendingTime"))
                .setMsgSeqNum(Integer.parseInt(getField(headerField, "MsgSeqNum")))
                .setBodyLength(Integer.parseInt(getField(headerField, "BodyLength")))
                .setMsgType(getField(headerField, "MsgType"))
                .setTargetCompId(getField(headerField, "TargetCompId"))
                .build();

        ExecutionReport executionReport = ExecutionReport.newBuilder()
                .setSecurityId(getField(message, "SecurityID"))
                .setSecurityIdSource(getField(message, "SecurityIDSource"))
                .setOrdType(getField(message, "OrdType"))
                .setAccountType(Integer.parseInt(getField(message, "AccountType")))
                .setOrderCapacity(getField(message, "OrderCapacity"))
                .setClOrdId(getField(message, "ClOrdID"))
                .setOrderQty(Float.parseFloat(getField(message, "OrderQty")))
                .setLeavesQty(Float.parseFloat(getField(message, "LeavesQty")))
                .setSide(getField(message, "Side"))
                .setCumQty(Float.parseFloat(getField(message, "CumQty")))
                .setExecType(getField(message, "ExecType"))
                .setOrdStatus(getField(message, "OrdStatus"))
                .setExecId(getField(message, "ExecID"))
                .setPrice(Float.parseFloat(getField(message, "Price")))
                .setOrderId(getField(message, "OrderID"))
                .setText(getField(message, "Text"))
                .setTimeInForce(getField(message, "TimeInForce"))
                .setTransactTime(getField(message, "TransactTime"))
                .setHeader(header)
                .setLastPx(Float.parseFloat(getField(message, "LastPx")))
                .build();

        return ResponseMessageTyped.newBuilder().setExecutionReport(executionReport).build();
    }

    private ResponseMessageTyped createQuoteStatusReport(Message message) {
        QuoteStatusReport quoteStatusReport = QuoteStatusReport.newBuilder()
                .setQuoteId(getField(message, "QuoteID"))
                .setQuoteStatus(Integer.parseInt(getField(message, "QuoteStatus")))
                .setSecurityId(getField(message, "SecurityID"))
                .setSecurityIdSource(getField(message, "SecurityIDSource"))
                .setSymbol(getField(message, "Symbol"))
                .setText(getField(message, "Text")).build();

        return ResponseMessageTyped.newBuilder().setQuoteStatusReport(quoteStatusReport).build();
    }

    private ResponseMessageTyped createQuote(Message message) {
        Quote quote = Quote.newBuilder()
                .addAllNoQuoteQualifiers(createQuoteQualifier(message))
                .setOfferPx(Float.parseFloat(getField(message, "OfferPx")))
                .setOfferSize(Float.parseFloat(getField(message, "OfferSize")))
                .setQuoteId(getField(message, "QuoteID"))
                .setSymbol(getField(message, "Symbol"))
                .setSecurityIdSource(getField(message, "SecurityIDSource"))
                .setBidSize(getField(message, "BidSize"))
                .setBidPx(Float.parseFloat(getField(message, "BidPx")))
                .setSecurityId(getField(message, "SecurityID"))
                .setQuoteType(Integer.parseInt(getField(message, "QuoteType"))).build();

        return ResponseMessageTyped.newBuilder().setQuote(quote).build();
    }

    private ResponseMessageTyped createQuoteAck(Message message) {
        QuoteAck quoteAck = QuoteAck.newBuilder()
                .setRfqid(getField(message, "RFQID"))
                .setText(getField(message, "Text")).build();

        return ResponseMessageTyped.newBuilder().setQuoteAck(quoteAck).build();
    }

    private ResponseMessageTyped createOrderMassCancelReport(Message message) {
        OrderMassCancelReport orderMassCancelReport = OrderMassCancelReport.newBuilder()
                .setClOrdId(getField(message, "ClOrdID"))
                .setText(getField(message, "Text")).build();

        return ResponseMessageTyped.newBuilder().setOrderMassCancelReport(orderMassCancelReport).build();
    }

    private ResponseMessageTyped createMassQuoteAcknowledgement(Message message) {
        MassQuoteAcknowledgement massQuoteAcknowledgement = MassQuoteAcknowledgement.newBuilder()
                .setQuoteId(getField(message, "QuoteID"))
                .setText(getField(message, "Text")).build();

        return ResponseMessageTyped.newBuilder().setMassQuoteAcknowledgement(massQuoteAcknowledgement).build();
    }

    private String getField(Message message, String key) {
        try {
            return message.getFieldsMap().get(key).getSimpleValue();
        } catch (NullPointerException npe) {
            LOGGER.error("There is no {} field in the message = {}", key, TextFormat.shortDebugString(message));
            throw npe;
        }
    }

    private List<Quote.QuoteQualifier> createQuoteQualifier(Message message) {
        List<Quote.QuoteQualifier> quoteQualifiers = new ArrayList<>();
        for (Value value : message.getFieldsMap().get("NoQuoteQualifiers").getListValue().getValuesList()) {
            quoteQualifiers.add(Quote.QuoteQualifier.newBuilder()
                    .setQuoteQualifier(getField(value.getMessageValue(), "QuoteQualifier")).build());
        }
        return quoteQualifiers;
    }
}

File: src/main/kotlin/com/exactpro/th2/act/convertors/ConvertorsResponse.kt
/*
 * Copyright 2022-2022 Exactpro (Exactpro Systems Limited)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exactpro.th2.act.convertors

import com.exactpro.th2.act.grpc.*
import com.exactpro.th2.act.grpc.NoPartyIDs
import com.exactpro.th2.act.grpc.Quote.QuoteQualifier
import com.exactpro.th2.common.grpc.Message
import com.exactpro.th2.common.grpc.Value
import com.google.protobuf.TextFormat
import org.slf4j.LoggerFactory


class ConvertorsResponse {
    fun createResponseMessage(message: Message): ResponseMessageTyped {
        val responseMessageTyped: ResponseMessageTyped =
            when (val messageType = message.metadata.messageType) {
                "QuoteStatusReport" -> createQuoteStatusReport(message)
                "Quote" -> createQuote(message)
                "QuoteAck" -> createQuoteAck(message)
                "OrderMassCancelReport" -> createOrderMassCancelReport(message)
                "MassQuoteAcknowledgement" -> createMassQuoteAcknowledgement(message)
                else -> throw IllegalArgumentException("Unsupported request message type $messageType")
            }
        return responseMessageTyped
    }

    /*private fun createBusinessMessageReject(message: Message): ResponseMessageTyped {
        val businessMessageReject = BusinessMessageReject.newBuilder().apply {
            refMsgType = field(message, "RefMsgType")
            businessRejectReason = field(message, "BusinessRejectReason").toInt()
            businessRejectRefId = field(message, "BusinessRejectRefID")
            refSeqNum = field(message, "RefSeqNum").toInt()
            text = field(message, "Text")
        }

        return ResponseMessageTyped.newBuilder()
            .setBusinessMessageReject(businessMessageReject).build()
    }

    private fun createExecutionReport(message: Message): ResponseMessageTyped {
        val executionReport = ExecutionReport.newBuilder().apply {
            securityId = field(message, "SecurityID")
            securityIdSource = field(message, "SecurityIDSource")
            ordType = field(message, "OrdType")
            accountType = field(message, "AccountType").toInt()
            orderCapacity = field(message, "OrderCapacity")
            clOrdId = field(message, "ClOrdID")
            orderQty = field(message, "OrderQty").toFloat()
            leavesQty = field(message, "LeavesQty").toFloat()
            side = field(message, "Side")
            cumQty = field(message, "CumQty").toFloat()
            execType = field(message, "ExecType")
            ordStatus = field(message, "OrdStatus")
            execId = field(message, "ExecID")
            price = field(message, "Price").toFloat()
            orderId = field(message, "OrderID")
            text = field(message, "Text")
            timeInForce = field(message, "TimeInForce")
            transactTime = field(message, "TransactTime")
            lastPx = field(message, "LastPx").toFloat()
        }

        val headerFields = message.fieldsMap["Header"]
        if (headerFields != null) {
            val headerField = headerFields.messageValue

            val header = Header.newBuilder().apply {
                beginString = field(headerField, "BeginString")
                senderCompId = field(headerField, "SenderCompId")
                sendingTime = field(headerField, "SendingTime")
                msgSeqNum = field(headerField, "MsgSeqNum").toInt()
                bodyLength = field(headerField, "BodyLength").toInt()
                msgType = field(headerField, "MsgType")
                targetCompId = field(headerField, "TargetCompId")
            }

            executionReport.header = header.build()
        }

        return ResponseMessageTyped.newBuilder()
            .setExecutionReport(executionReport).build()
    }*/

    private fun createQuoteStatusReport(message: Message): ResponseMessageTyped {
        val quoteStatusReport = QuoteStatusReport.newBuilder().apply {
            quoteId = field(message, "QuoteID")
            quoteStatus = field(message, "QuoteStatus").toInt()
            securityId = field(message, "SecurityID")
            securityIdSource = field(message, "SecurityIDSource")
            symbol = field(message, "Symbol")
            text = field(message, "Text")
        }

        return ResponseMessageTyped.newBuilder()
            .setQuoteStatusReport(quoteStatusReport).build()
    }

    private fun createQuote(message: Message): ResponseMessageTyped {
        val quote = Quote.newBuilder().apply {
            addAllNoQuoteQualifiers(createQuoteQualifier(message))
            offerPx = field(message, "OfferPx").toFloat()
            offerSize = field(message, "OfferSize").toFloat()
            quoteId = field(message, "QuoteID")
            symbol = field(message, "Symbol")
            securityIdSource = field(message, "SecurityIDSource")
            bidSize = field(message, "BidSize")
            bidPx = field(message, "BidPx").toFloat()
            securityId = field(message, "SecurityID")
            addAllNoPartyIds(createNoPartyIDs(message))
            quoteType = field(message, "QuoteType").toInt()
        }
        return ResponseMessageTyped.newBuilder()
            .setQuote(quote).build()
    }

    private fun createQuoteAck(message: Message): ResponseMessageTyped {
        val quoteAck = QuoteAck.newBuilder().apply {
            rfqid = field(message, "RFQID")
            text = field(message, "Text")
        }
        return ResponseMessageTyped.newBuilder()
            .setQuoteAck(quoteAck).build()
    }

    private fun createOrderMassCancelReport(message: Message): ResponseMessageTyped {
        val orderMassCancelReport = OrderMassCancelReport.newBuilder().apply {
            clOrdId = field(message, "ClOrdID")
            text = field(message, "Text")
        }
        return ResponseMessageTyped.newBuilder()
            .setOrderMassCancelReport(orderMassCancelReport).build()
    }

    private fun createMassQuoteAcknowledgement(message: Message): ResponseMessageTyped {
        val massQuoteAcknowledgement = MassQuoteAcknowledgement.newBuilder().apply {
            quoteId = field(message, "QuoteID")
            text = field(message, "Text")
        }

        return ResponseMessageTyped.newBuilder()
            .setMassQuoteAcknowledgement(massQuoteAcknowledgement).build()
    }

    private fun field(message: Message, key: String): String =
        try {
            message.fieldsMap[key]!!.simpleValue
        } catch (npe: NullPointerException) {
            LOGGER.error("There is no {} field in the message = {}", key, TextFormat.shortDebugString(message))
            throw npe
        }

    private fun createQuoteQualifier(message: Message): List<QuoteQualifier> {
        val quoteQualifiers: MutableList<QuoteQualifier> = ArrayList()
        val noQuoteQualifiers = message.fieldsMap["NoQuoteQualifiers"]
        if(noQuoteQualifiers != null) {
            for (value in noQuoteQualifiers.listValue.valuesList) {
                quoteQualifiers.add(
                    QuoteQualifier.newBuilder()
                        .setQuoteQualifier(field(value.messageValue, "QuoteQualifier")).build()
                )
            }
        }
        return quoteQualifiers
    }

    fun createNoPartyIDs(message: Message): List<NoPartyIDs>{
        val msgNoPartyIDs: List<Value> = message.fieldsMap["NoPartyIDs"]!!.listValue.valuesList
        val noPartyIDs: MutableList<NoPartyIDs> = ArrayList()
        for (value in msgNoPartyIDs) {
            val msg: Message = value.messageValue
            noPartyIDs.add(
                NoPartyIDs.newBuilder()
                    .setPartyId(field(msg, "PartyID"))
                    .setPartyIdSource(field(msg, "PartyIDSource"))
                    .setPartyRole(field(msg, "PartyRole").toInt()).build()
            )
        }
        return noPartyIDs
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(ConvertorsResponse::class.java)
    }
}